### ISSUE TITLE
probable socket leak in C*KVS downed host pool re-add checker

### ISSUE THREAD

**Issue by clockfort** [palantir/atlasdb#1654]
Found an issue at a customer site that would be most easily explained by the titular situation.

Symptoms:
- Atlas-using JVM has normal jstack
- Cassandra server JVM has bonkers 32K+ idle thrift receiver threads waiting on input
- Examination of OS-level network stack on both sides reveals active open TCP/SSL connections on both boxes going from atlas to C* server
- Happened during a time customer incorrectly added a Cassandra node to cluster and Atlas (correctly) blacklisted it after determining the host to be continually unreliable

It's probable the issue can be teased out with a thorough examination of the codepath around dealing with dead Cassandra nodes and how we re-check when we attempt to add the host back into the fold if it starts behaving again.

I will get around to looking at this and making a patch, just filing a github issue to keep myself honest / reminded.

**Comment by jeremyk-91** [comment]
@clockfort Thanks. Do you plan on completing this this week? In any case, we'll be happy to review the PR once it's ready.

**Comment by jboreiko** [comment]
@clockfort pointed out this is likely associated with a bug in the thrift client https://datastax-oss.atlassian.net/browse/JAVA-1072

**Comment by rhero** [comment]
@clockfort given that we merged the dependency bump, do you want to confirm that we've fixed the issue at the customer?

**Comment by clockfort** [comment]
I haven't merged the change into $downstreamProject yet, I will today or tomorrow. It's been extremely aggravating to bump $downstreamProject onto this version because it has a ton of newer library version conflicts to work through, even after the extensive shading I did to put this in.

I could verify this actually fixes the bug at any time, though, if you'd like, because it should appear in OSS Atlas itself as well.

**Comment by rhero** [comment]
I don't think it's super high priority to confirm - I just saw #1753 and wanted to ping this in case you've already done it.

**Comment by hsaraogi** [comment]
@clockfort are you planning to get to this?

**Comment by clockfort** [comment]
yep, I'm temporarily fresh out of releases and P0s to do, so I can work on this now.

**Comment by jboreiko** [comment]
https://github.com/palantir/atlasdb/pull/1896
